pkg/cli/dockerx/docker.go

```diff
@@ -1,38 +1,123 @@
 package dockerx
 
 import (
+	"bufio"
 	"context"
+	"fmt"
+	"io"
 	"os"
+	"os/signal"
 	"path"
+	"syscall"
 
+	"github.com/docker/docker/api/types"
 	"github.com/docker/docker/api/types/container"
 	"github.com/docker/docker/api/types/filters"
+	"github.com/docker/docker/api/types/network"
 	"github.com/docker/docker/client"
-	"github.com/magefile/mage/sh"
+	"github.com/docker/go-connections/nat"
+	v1 "github.com/opencontainers/image-spec/specs-go/v1"
 )
 
-const (
-	docker string = "docker"
-)
+func PolicyRoot() string {
+	const defaultPolicyRoot = ".policy"
 
-var (
-	DockerRun = sh.RunCmd(docker)
-	DockerOut = sh.OutCmd(docker)
-)
+	policyRoot := os.Getenv("POLICY_FILE_STORE_ROOT")
+	if policyRoot == "" {
+		home, err := os.UserHomeDir()
+		if err != nil {
+			return ""
+		}
 
-func DockerV(args ...string) error {
-	return sh.RunV(docker, args...)
+		return path.Join(home, defaultPolicyRoot)
+	}
+	return policyRoot
+}
+
+type DockerClient struct {
+	ctx context.Context
+	cli *client.Client
 }
 
-func IsRunning(name string) (bool, error) {
+func New() (*DockerClient, error) {
 	cli, err := client.NewClientWithOpts(
 		client.FromEnv,
 		client.WithAPIVersionNegotiation())
 	if err != nil {
-		return false, err
+		return nil, err
+	}
+
+	return &DockerClient{
+		ctx: context.Background(),
+		cli: cli,
+	}, nil
+}
+
+// Pull container image.
+func (dc *DockerClient) Pull(image, platform string) error {
+	out, err := dc.cli.ImagePull(dc.ctx, image, types.ImagePullOptions{
+		Platform: platform,
+	})
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+
+	_, _ = io.Copy(io.Discard, out)
+
+	return nil
+}
+
+// Remove container image as image-name:tag.
+func (dc *DockerClient) RemoveImage(image string) error {
+	images, err := dc.cli.ImageList(dc.ctx, types.ImageListOptions{
+		Filters: filters.NewArgs(
+			filters.KeyValuePair{
+				Key: "reference", Value: image,
+			},
+		),
+	})
+	if err != nil {
+		return err
+	}
+
+	for i := 0; i < len(images); i++ {
+		_, err := dc.cli.ImageRemove(dc.ctx, images[i].ID, types.ImageRemoveOptions{})
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Stop container instance with `name`.
+func (dc *DockerClient) Stop(name string) error {
+	containers, err := dc.cli.ContainerList(dc.ctx, container.ListOptions{
```

Comment:
Can we use the DockerClient IsRunning to check before calling dc.cli.ContainerStop ?

Reply:
That is not appropriate at this level, which provides the building block. the `topaz stop` command handler does that, calling 
```
if c.CheckRunStatus(cmd.ContainerName, cc.StatusNotRunning) {
		return nil
	}
```
before calling dockerx.Stop()

https://github.com/aserto-dev/topaz/blob/93b515ce3ca5becd5c4b859d94bc0671a2835a49/pkg/cli/cmd/stop.go#L17